### DISCUSSION TITLE
Deployment of Fips provider onto the target machine

### DISCUSSION THREAD

**Discussion by nasir-professional** [openssl/openssl#25708] [Q&A] [answered]
I am using openssl-3.0.9 version which is Fips validated to build and install openssl and Fips provider on Windows platform and using Perl, NASM and Visual Studio for it.

It produces the required libraries and the configuration files that openssl and fips provider needs to work. My application reads "openssl.cnf" configuration programmatically and loads the Fips provider successfully. 

**_OpenSSL documentation says that The FIPS module must have the self tests run, and the FIPS module config file output generated on every machine that it is to be used on. For OpenSSL 3.0, you must not copy the FIPS module config file output data from one machine to another._**

I have got a question about the deployment of my solution onto the customer's target machine to help understand what does it need to ensure that it follows the Fips compliance requirements. 

Q-1: Does fips provider need to be installed on each target machine to have fips.dll or if I can use the one which was produced in my environment when building and installing the openssl and fips provider?

Q-2: Could the fips.dll file produced in my environment be used on each target machine to generate fipsmodule.cnf file or would it need to use the one produced on the target machine as a result of installing fips provider on to it?

Q-3: As for the self-tests, correct me if I am wrong that fips module supports running the self-tests on load. It can be achieved when executing openssl fipsinstall command with -self_test_onload parameter when generating fipsmodule.cnf file. install-mac and install-status should not exist in the .cnf file. In this way then we do not need to run the self-tests separately on the machine as they will run on each load automatically, right?

**Comment by paulidale** [comment, marked as answer]
1. Yes, you only need to build fips.dll once.
2. No, you'd be best running the fipsinstall command to create the fipsmodule.cnf.  fips.dll won't load without fipsmodule.cnf being present.
3. Yes, you wouldn't need to run the self tests on install if you pick this option.  However, the security policy doesn't allow this option.  You would meet the _spirit_ of the FIPS rules but not the _letter_ of them.  Thus, you'd be non-compliant.  Others _have_ chosen this option but it is a risk if you are audited.  The FIPS 140-3 provider, when it is released, will not support the self test on install option and will only run self tests on load.

**Reply by t8m** [reply]
> 3\. The FIPS 140-3 provider, when it is released

I would just make this more precise - the FIPS 140-3 provider was released but it is not validated yet.

**Reply by nasir-professional** [reply]
Thanks for the answers.

Just for the sake of double checking that I understood it correctly.

Q1: Fips.dll produced in my build environment (separate machine) can be deployed/used onto the customer's target machine, right?

Q2: openssl fipsinstall command will need to be run on each customer's target machine. It will do two things; 1) runs the FIPS module self tests 2) generates fipsmodule.cnf file. This way It will fulfil the criteria of security policy. Below is an example command I am just quoting here for reference purpose. 

openssl fipsinstall -module "C:\openssl-3.0.9\Delivery\x64\fips.dll" -out "C:\openssl-3.0.9\Delivery\x64\fipsmodule.cnf"

Q3: I notice that contents of fipsmodule.cnf file are same when it is generated running the example command in Q2 on each target machine. Is it expected?

Q4: Is there any requirement to build openssl Fips validated version 3.0.9 on any specific windows operating system version? For the Windows platform, the tested configurations for Fips compliance specifies Windows 10 (x64) in CMVP Certificate #4282. Since this is part of the validated configurations so just wondering if it is the recommended operating system for windows platform or If I could use any other version as well to build/install openssl fips provider? I have been able to build it successfully on Windows 11 and Windows Server 2019 using native tools command prompt of Visual Studio 2017 and 2022 respectively. Are there any guidelines on this?

**Reply by paulidale** [reply]
To get definitive answers to your questions, engage with a certified FIPS laboratory.  They are the only folks who can give you official answers.
